pkg/ebpf/ingress_node_firewall_loader.go

```diff
@@ -55,43 +59,295 @@ func NewIngNodeFwController() (*IngNodeFwController, error) {
+		return err
+	}
+
+	// Purge all stale keys from the eBPF map.
+	if err := infc.purgeKeys(staleKeys); err != nil {
+		klog.Infof("Purge keys operation encountered issues, err: %q", err)
 	}
-	klog.Infof("Ingress node firewall map Info: %+v with FD %s", info, objs.BpfMaps.IngressNodeFirewallTableMap.String())
 
-	if err := infc.makeIngressFwRulesMap(ingFireWallConfig, isDelete, ifId); err != nil {
-		return fmt.Errorf("Failed to create map firewall rules: %v on if %d", err, ifId)
+	// Add/update all keys. This is an idempotent action and non-existing keys are added whereas existing keys
+	// are updated.
+	if err := infc.addOrUpdateRules(ebpfKeyToRules); err != nil {
+		return err
 	}
 
+	// Generate ingress node few events.
 	infc.ingressNodeFwEvents()
 
 	return nil
 }
 
+// addOrUpdateRules is a small method containing this limited set of functionality to facilitate unit testing of
+// this part of code.
+func (infc *IngNodeFwController) addOrUpdateRules(ebpfKeyToRules map[BpfLpmIpKeySt]BpfRulesValSt) error {
+	for ebpfKey, ebpfRules := range ebpfKeyToRules {
+		log.Printf("Adding or updating ingress firewall rules for key %v", ebpfKey)
+		if err := infc.objs.BpfMaps.IngressNodeFirewallTableMap.Update(ebpfKey, ebpfRules, ebpf.UpdateAny); err != nil {
+			return fmt.Errorf("Failed Adding/Updating ingress firewall rules: %v", err)
+		}
+	}
+	return nil
+}
+
+// GetStatisticsMap returns the statistics map of the object.
 func (infc *IngNodeFwController) GetStatisticsMap() *ebpf.Map {
 	return infc.objs.IngressNodeFirewallStatisticsMap
 }
 
-// makeIngressFwRulesMap convert IngressNodeFirewallRules into eBPF format which matched what the
-// kerenl hook will be using.
-func (infc *IngNodeFwController) makeIngressFwRulesMap(ingFirewallConfig ingressnodefwiov1alpha1.IngressNodeFirewallRules,
-	isDelete bool, ifId uint32) error {
+// IngressNodeFwAttach attaches the eBPF program to a given list of interfaces and pins them to different pinDirs.
+// For each provided interface name:
+// i)   Look up the network interface by name.
+// ii)  Attach the program to the interface.
+// iii) Pin the XDP program.
+func (infc *IngNodeFwController) IngressNodeFwAttach(ifacesName ...string) error {
 	objs := infc.objs
-	rules := BpfRulesValSt{}
+	for _, ifaceName := range ifacesName {
+		if _, ok := infc.links[ifaceName]; ok {
+			klog.Infof("Interface %s is already attached and managed, skipping", ifaceName)
+			return nil
+		}
+		// Look up the network interface by name.
+		iface, err := net.InterfaceByName(ifaceName)
+		if err != nil {
+			return fmt.Errorf("lookup network iface %q: %s", ifaceName, err)
+		}
+
+		// Attach the program.
+		l, err := link.AttachXDP(link.XDPOptions{
+			Program:   objs.IngressNodeFirewallProcess,
+			Interface: iface.Index,
+		})
+		if err != nil {
+			if strings.Contains(err.Error(), xdpEBUSYErr) {
+				return fmt.Errorf("Interface %s is already attached", ifaceName)
```

Comment:
EBUSY isn't an error right ? unless we aren't suppose to get this condition now in that case we need to remove teh special check for EBUSY and just error if err != nil

Reply:
I removed that check as with the new "scan the folder and create link objects from pins", the dual attach doesn't happen any more.

On the other hand, I used to run into race conditions between detach/attach operations during the unit tests so I'm trying to catch those here:
~~~
  204 func (e *ebpfSingleton) attachNewInterfaces(ifaceIngressRules map[string][]v1alpha1.IngressNodeFirewallRules) error {   
  205     for intf := range ifaceIngressRules {                                                                               
  206         // First, check if the interface name is valid.                                                                 
  207         if !isValidInterfaceNameAndState(intf) {                                                                        
  208             return fmt.Errorf("Fail to attach ingress node Firewall rules: invalid interface %s", intf)                 
  209         }                                                                                                               
  210                                                                                                                         
  211         // Then, check if the interface is already managed.                                                             
  212         if _, ok := e.managedInterfaces[intf]; !ok {                                                                    
  213             // Attach to the interfaces - in case the interface is already attached, retry.                             
  214             err := retry.OnError(                                                                                       
  215                 retry.DefaultRetry,                                                                                     
  216                 func(err error) bool {                                                                                  
  217                     return strings.Contains(err.Error(), "device or resource busy")                                     
  218                 },                                                                                                      
  219                 func() error {                                                                                          
  220                     e.log.Info("Attaching firewall interface", "intf", intf)                                            
  221                     if err := e.c.IngressNodeFwAttach(intf); err != nil {                                               
  222                         e.log.Error(err, "Fail to attach ingress firewall prog")                                        
  223                         return err                                                                                      
  224                     }                                                                                                   
  225                     e.managedInterfaces[intf] = struct{}{}                                                              
  226                     return nil                                                                                          
  227                 })                                                                                                      
  228             if err != nil {                                                                                             
  229                 return err                                                                                              
  230             }                                                                                                           
  231         }                                                                                                               
  232     }                                                                                                                   
  233     return nil                                                                                                          
  234 }       
~~~

But I'm having a hard time replicating that issue at the moment.